### ISSUE TITLE
[Rejected] SDL 0309 - Mitigating Android SDL service notifications issues

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#1051]
Hello SDL community,

The review of "SDL 0309 - Mitigating Android SDL service notifications issues"  begins now and runs through July 28, 2020.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0309-android-notifications.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/1051

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeygrover** [comment]
1\. Exit button on router service notificaiton

Allowing users to close the SDL Router service is more likely to bring up issues than it is going to solve. Having the ability to close the service when it is running properly is incorrect. The user can force close the app that started the notification and thus would have a better understanding that what they did is not something we protect for. Any adverse effects seen from this force close would be on them, not us. The only time I could see this has potentially useful is during the pre-connection phase where it is waiting for a connection. However, it should be known that even though this notification is displayed on the device, it doesn't mean that the service hasn't tried to remove it or is actually closed. This was discussed in a previous proposal, but essentially the Android OS is not honoring the notification removal properly so adding a close may have little to no effect.

2\. Code Snippets

>The implementation within the SDL broadcast receiver to avoid starting the router service is internal, hence it should be considered a development detail.

This is not an acceptable answer. As we've discussed in other proposals this feature is very complex and creating a wish list of requirements is not satisfactory for accepting a potential feature. For example, how does the broadcast receiver know that this private variable is set to false? Setting a static variable from one process into a class that is instantiated into another process does not work; each process contains its own set of static variables and they are not shared between processes. So I don't believe this feature will work as the author is intending.


3\. Developer set pending intent

I don't think this is the right or most intuitive place for this. The issue will be that any app can potentially host the router service so navigating to an app's specific settings to turn off SDL from this router service will be disingenuous to the user. As if even the user disables SDL in that app, another app will start the router service with the same message. In my opinion the better place for such a pending intent would be the app's own SDL Service notification. They have complete control over this notification and can alter it in anyway that fits their needs. Then there is the issue of app's changing what happens with this pending intent based on their choices. Without a standard or limitation they can control what happens and each click of the notification may result in a different experience for the user based on who is hosting the router service. This could be a simple code example in the documentation rather than adding confusing and an infinitely different UX to the library. 

4\. Turning off SDL via the router service
 
API to I am against adding the ability to disable the router service at runtime by an application. This adds yet ANOTHER state that to the router service's lifecycle that the author is continually adding to with other proposals. Also, this can be achieved by the developer already and with great success than adding a flag to check at runtime for every start.

Instead developers can add code to their Application class like the follow:

```java
    public void enableSDL(boolean enabled){
        PackageManager pm  = getPackageManager();
        enableComponent(pm, enabled, "com.sdl.hellosdlandroid", "com.sdl.hellosdlandroid.SdlRouterService");
        enableComponent(pm, enabled, "com.sdl.hellosdlandroid", "com.sdl.hellosdlandroid.SdlService");
        enableComponent(pm, enabled, "com.sdl.hellosdlandroid", "com.sdl.hellosdlandroid.SdlReceiver");
    }

    private void enableComponent(PackageManager pm, boolean enabled, String packageName, String componentClassName){
        ComponentName componentName = new ComponentName(packageName, componentClassName);
        int state = enabled ?  PackageManager.COMPONENT_ENABLED_STATE_ENABLED :  PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
        pm.setComponentEnabledSetting(componentName, state, PackageManager.DONT_KILL_APP);
    }
```

This would prevent other apps from even finding that apps router service. It would stop intents from being sent to the app and their SDL service would never get started. This would prevent the library from having to add another check and state to complicated an already complicated flow. This could be a simple code example in the documentation instead of extra code bloat in the library.


5\. Overall thoughts

I believe we have all recognized the ineptitude of the Android notification system as well as the foreground service requirements at this point. With every release it has been a challenge to fit our solution into their constantly changing requirements and new APIs. While we have followed their documentation and added safeguards, the issue of Android fragmentation and incorrect API behavior plagues this feature. As the project maintainer and someone actively working to solve this issue in the past, I would have appreciated some mention that this is not all due to what we are doing in the library and the blame is not solely on SDL. As we've progressed trying to address this and issues are raised, we as the project maintainer have asked time and time again for reproduction steps to no avail. 

I believe the title and problem statement of this proposal is also a bit off point with some of the requested changes that are bundled into this single proposal. This proposal should actually read "Allow users to disable SDL" because the solutions themselves don't really fix the issues with notifications. 



In conclusion, I believe most of the changes requested within this proposal to be an extra complication that can be solved by simply adding documentation. With proposals [SDL Device Listener](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0301-SDL-device-listener.md) and potentially [Enable OEM exclusive apps support](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0293-vehicle-type-filter.md) I find this proposal unnecessary as the notification issue should be solved meaning that mitigation is no longer required. All other changes requested can be addressed through documentation.

**Comment by theresalech** [comment]
The Steering Committee voted to reject this proposal, based on the comments provided by the PM [here](https://github.com/smartdevicelink/sdl_evolution/issues/1051#issuecomment-663657308).  As noted in the linked comment, the notification issue should be solved (meaning mitigation is no longer required) via SDL 0301 and SDL 0293, and other changes included in this proposal can be addressed through documentation.
